### ISSUE TITLE
Wrong file length and return code on arecord v1.2.7 (working fine on v1.2.6)

### ISSUE THREAD

**Issue by dolcini** [alsa-project/alsa-utils#163]
arecord returns a failure exit code and records a file with the wrong size in v1.2.7, while v1.2.6 is working fine.

My test is recording 3 seconds, 16bit, 44100Hz, mono, `arecord -D hw:2,0 -c 1 -f S16_LE -r 44100 -t wav -d 3 mic.wav`. Expected output file size is 264644 bytes = 3s * 2bytes (16bits) * 44100Hz * 1 + 44bytes (sizeof wav.hdr) and arecord should return a 0 exit code.

This is working fine with v1.2.6, with v1.2.7 the recorded file is 24bytes bigger than expected and arecord exit code is 1.

Test is executed on a 32bit arm (toradex apalis-imx6), with alsa-utils compiled with openembedded, http://cgit.openembedded.org/openembedded-core/tree/meta/recipes-multimedia/alsa running on a 5.19 Linux Kernel. No differences running with a 5.4 kernel. A similar issue was reproduced with a different audio codec on arm64, I do not have the logs at hand now, however.

This commit [`4b959a447aa1 ("aplay: Fix for arecord recording ghost data")`](https://github.com/alsa-project/alsa-utils/commit/4b959a447aa1e9ce8d0076fc59fa71ff15f025c3) from Amadeusz Sławiński <amadeuszx.slawinski@linux.intel.com> @amadeuszslawinski-intel is likely the root cause.


Working:

```
root@apalis-imx6-10551895:~# cat /proc/asound/cards 
 0 [DWHDMI         ]: dw-hdmi-ahb-aud - DW-HDMI
                      DW-HDMI rev 0x0a, irq 273
 1 [imxspdif       ]: imx-spdif - imx-spdif
                      imx-spdif
 2 [imx6qapalissgtl]: imx6q-apalis-sg - imx6q-apalis-sgtl5000
                      imx6q-apalis-sgtl5000

root@apalis-imx6-10551895:~# amixer -c 2 set "Capture Mux" "MIC_IN"                                                                                                                                                                                                                       
Simple mixer control 'Capture Mux',0
  Capabilities: enum
  Items: 'MIC_IN' 'LINE_IN'
  Item0: 'MIC_IN'
root@apalis-imx6-10551895:~# echo $?
0

root@apalis-imx6-10551895:~# arecord -D hw:2,0 -c 1 -f S16_LE -r 44100 -t wav -d 3 mic.wav
Recording WAVE 'mic.wav' : Signed 16 bit Little Endian, Rate 44100 Hz, Mono
root@apalis-imx6-10551895:~# echo $?
0

root@apalis-imx6-10551895:~# ls -l mic.wav 
-rw-r--r-- 1 root root 264644 Aug  3 14:44 mic.wav

root@apalis-imx6-10551895:~# arecord --version
arecord: version 1.2.6 by Jaroslav Kysela <perex@perex.cz>

```

Broken
```
root@apalis-imx6-10551895:~# amixer -c 2 set "Capture Mux" "MIC_IN"     
Simple mixer control 'Capture Mux',0
  Capabilities: enum
  Items: 'MIC_IN' 'LINE_IN'
  Item0: 'MIC_IN'
root@apalis-imx6-10551895:~#  echo $?
0

root@apalis-imx6-10551895:~#  arecord -D hw:2,0 -c 1 -f S16_LE -r 44100 -t wav -d 3 mic.wav
Recording WAVE 'mic.wav' : Signed 16 bit Little Endian, Rate 44100 Hz, Mono
root@apalis-imx6-10551895:~# echo $?
1

root@apalis-imx6-10551895:~# ls -l mic.wav 
-rw-r--r-- 1 root root 264668 Aug  3 14:55 mic.wav

root@apalis-imx6-10551895:~# arecord --version
arecord: version 1.2.7 by Jaroslav Kysela <perex@perex.cz>

```

**Comment by dolcini** [comment]
I was able to try also on my x86-64 PC and the issue is just the same and it's very easy to reproduce. Reverting `4b959a447aa1 ("aplay: Fix for arecord recording ghost data")` fixes it. @amadeuszslawinski-intel can you help?

**Comment by perexg** [comment]
Thanks. Fixed in 6ee086363a62c9692787860c1edf4f618ad1575a .
